### PR TITLE
Make kubeadm config imageRepository can implement coredns

### PR DESCRIPTION
**Pull request by nashasha1**
**What this PR does / why we need it**:
This PR make the kubeadm's config `imageRepository` can implement on `coredns`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
I know coredns images configuration for the Deployment are using index.docker.io as a registry, and I think we should support overwriting the registry just like other repositories. There are occasions or places where access to grc.io or index.docker.io registry is restricted, therefore there should be a way to set a mirror as a fallback. The only downside from this implementation would be to add an external process to sync coredns images to the default registry. I am aware that this could add complexity to the process, and I am open to discuss further a solution to this problem, including changing this pull-request to satisfy both needs.

**Release note**:

```release-note
      Kubeadm's coredns deployment configuration using imageRepository just like other images
```

### CODE REVIEW

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 1]
```diff
@@ -141,9 +141,11 @@ func coreDNSAddon(cfg *kubeadmapi.MasterConfiguration, client clientset.Interfac
 	}
 
 	// Get the config file for CoreDNS
-	coreDNSConfigMapBytes, err := kubeadmutil.ParseTemplate(CoreDNSConfigMap, struct{ DNSDomain, ServiceCIDR string }{
-		ServiceCIDR: cfg.Networking.ServiceSubnet,
-		DNSDomain:   cfg.Networking.DNSDomain,
+	coreDNSConfigMapBytes, err := kubeadmutil.ParseTemplate(CoreDNSConfigMap,
+		struct{ ImageRepository, DNSDomain, ServiceCIDR string }{
+			ImageRepository: cfg.ImageRepository,
```
Comment (dixudx) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: Use `cfg.GetControlPlaneImageRepository()` instead.

**Review by dixudx — CHANGES_REQUESTED** [review]
Please run `./hack/update-gofmt.sh`.

**Comment by nashasha1** [comment]
@dixudx   Done. Thank you for your help.

**Comment by fturib** [comment]
QUESTION: is that change compatible with BETA of coredns, and the fact that, before Final integration,  we get the image from coredns repository. (coredns/coredns ...)

Will the change with cfg.GetControlPlaneImageRepository() provide the same repository ?

Should we wait after BETA ?

**Comment by johnbelamaric** [comment]
Right now CoreDNS is pulled from Docker hub, and I think with this change it will start tying to pull it from gcr.io by default - and we don't have an image in there yet. As I understand it, there is currently an effort ongoing that will enable the building of all necessary containers from source and populating them into gcr.io, but I don't know the timeframe of that effort. Certainly when CoreDNS goes to *GA*, it will need to be sourced from gcr.io, but it isn't today.

In other words, I think this PR will break the default case, unless we have a process to get new CoreDNS images into gcr.io. @bowei PTAL

**Comment by nashasha1** [comment]
ping @bowei

File: cmd/kubeadm/app/phases/addons/dns/manifests.go [hunk 2]
```diff
@@ -257,7 +257,7 @@ spec:
               topologyKey: kubernetes.io/hostname
       containers:
       - name: coredns
-        image: coredns/coredns:{{ .Version }}
+        image: {{ .ImageRepository }}/coredns:{{ .Version }}
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/addons/dns/manifests.go]: this will change the default to gcr.io... why would we want that?

Comment (nashasha1) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: make the kubeadm's config `imageRepository` can implement on coredns.

Comment (nashasha1) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: @timothysc

Comment (johnbelamaric) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: We can't switch to gcr.io now. That will have to happen before GA but right now that will just cause a failure since there is no CoreDNS image there yet.

Comment (nashasha1) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: @johnbelamaric , Thanks for your reply. As I wrote in **Special notes for your reviewer**, the only problem for this PR is that we need a progress to sync the coreDNS image from `docker hub` to `gcr.io`.  I prefer make `kubeadm` to support overwrite the coreDNS's registry, even it is not GA.

Comment (johnbelamaric) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: From my past discussions with the team, the gcr.io piece will require a fork owned by k8s so that the Kubernetes community has control over the actual image delivered. We should bring this up in the sig-network meeting. The plan was to have that in place by 1.11.

Comment (johnbelamaric) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: But if there is another process we can use to populate grc.io, then this can work.

Comment (nashasha1) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: Thank you for your explanation. @johnbelamaric ! You mean the team has discussed about this question and decided to sync coreDNS images to grc.io in `Kubernetes v1.11`?   

Do we have the other way to get new coreDNS images into grc.io?  @timothysc .

Shall we create some mechanisms to handle this situation(break `kubeadm's` behavior) ?

Comment (timothysc) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: > Do we have the other way to get new coreDNS images into grc.io? @timothysc . 

Only a googler can do this /cc @roberthbailey

Comment (roberthbailey) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: /cc @bowei - is someone from sig-networking shepherding coreDNS images into gcr.io?

Comment (fturib) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: The transfer of CoreDNS images into grc.io will be part of the finalization for CoreDNS GA.
sig-networking asked to prepare a KEP for this phase.

I am currently working on its preparation. I expect this document to be available next week.
it will be located here : https://github.com/kubernetes/community/tree/master/keps/sig-network

I will reference this PR in the part of the moving generation under control of K8s and hosting the image in gcr.io.

**Comment by agau4779** [comment]
oops, wrong PR
/remove kind bug
/remove priority critical-urgent

**Comment by fturib** [comment]
The KEP for CoreDNS is out here : https://github.com/kubernetes/community/pull/1956
It includes a link to this PR when it comes to have an image available in gcr.io

**Comment by timothysc** [comment]
This requires some googler to push the image.  
/cc @ixdy @wojtek-t @kubernetes/sig-release-pr-reviews  I need a google-r that can push and verify this image

**Comment by timothysc** [comment]
/assign @thockin  

Someone at google needs todo the push before we an change this.

**Comment by thockin** [comment]
The discussion about hosting images etc is still ongoing, and I am not sure that the image will end up in our k8s GCR

**Comment by thockin** [comment]
working on how to push a manifest list

**Comment by fturib** [comment]
@thockin : Are the images now pushed to "gcr.io". If yes, i guess we can go ahead with approval and merge on this PR. Thanks!

**Comment by fturib** [comment]
@rajansandeep - wanted to be sure you are aware of this PR.

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 3]
```diff
@@ -169,12 +169,14 @@ func coreDNSAddon(cfg *kubeadmapi.MasterConfiguration, client clientset.Interfac
...
 	// Get the config file for CoreDNS
-	coreDNSConfigMapBytes, err := kubeadmutil.ParseTemplate(CoreDNSConfigMap, struct{ DNSDomain, UpstreamNameserver, Federation, StubDomain string }{
-		DNSDomain:          coreDNSDomain,
-		UpstreamNameserver: upstreamNameserver,
-		Federation:         federations,
-		StubDomain:         stubDomain,
-	})
+	coreDNSConfigMapBytes, err := kubeadmutil.ParseTemplate(CoreDNSConfigMap,
+		struct{ ImageRepository, DNSDomain, UpstreamNameserver, Federation, StubDomain string }{
+			ImageRepository:    cfg.ImageRepository,
```
Comment (rajansandeep) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: This should be under the `coreDNSDeploymentBytes`  and not here.

```
coreDNSDeploymentBytes, err := kubeadmutil.ParseTemplate(dnsDeploymentBytes, struct{ ImageRepository, MasterTaintKey, Version string }{
		ImageRepository: cfg.ImageRepository,
		MasterTaintKey:  kubeadmconstants.LabelNodeRoleMaster,
		Version:         GetDNSVersion(k8sVersion, kubeadmconstants.CoreDNS),
	})
```

**Review by rajansandeep — CHANGES_REQUESTED** [review]

**Comment by thockin** [comment]
Images are pushed to k8s.gcr.io now

**Comment by fturib** [comment]
@rajansandeep : as there is no reply from @nashasha1 and we need this change to be checked in today, before the code freeze, can you propose another PR (with same change) asap, so we can merge today ?
Let me know. thanks.

**Comment by rajansandeep** [comment]
@fturib I've opened #64775 with the same changes as here to ensure these changes get in before code freeze.

**Comment by roberthbailey** [comment]
Closing in favor of #64775.
